src/lib/Simplify.hs

```diff
@@ -367,18 +350,62 @@ simplifyAtomAndInline atom = confuseGHC >>= \_ -> case atom of
             LetBound (DeclBinding _ _ (Atom x)) -> dropSubst $ simplifyAtomAndInline x
             _ -> return $ Var v
 
-emitSpecializationName
-  :: Mut o => AtomName o -> [Atom o] -> SimplifyM i o (AtomName o)
-emitSpecializationName fname args = do
-  let specialization = AppSpecialization fname args
-  querySpecializationCache specialization >>= \case
-    Just specializedName -> return specializedName
+determineSpecializationSpec
+  :: EnvReader m => AtomName n -> [Atom n] -> m n (SpecializationSpec n, [Atom n])
+determineSpecializationSpec fname staticArgs = do
+  lookupAtomName fname >>= \case
+    TopFunBound (NaryPiType bs _ _) _ -> do
+      (PairB staticArgBinders _) <- return $
+        splitNestAt (length staticArgs) (nonEmptyToNest bs)
+      (ab, extraArgs) <- liftEnvReaderM $ generalizeDataComponentsNest staticArgBinders staticArgs
+      return (AppSpecialization (Var fname) ab, extraArgs)
+    _ -> error "should only be specializing top functions"
+
+type Generalized (e::E) (n::S) = (Abs (Nest Binder) e n, [Atom n])
+
+generalizeDataComponentsNest
+  :: Nest PiBinder n l
+  -> [Atom n]
+  -> EnvReaderM n (Generalized (ListE Atom) n)
+generalizeDataComponentsNest Empty [] = return (Abs Empty (ListE []), [])
+generalizeDataComponentsNest (Nest b bs) (x:xs) = do
+  (ab, newArgs) <- case binderType b of
+    TyKind -> generalizeType x
+    DictTy dTy -> do
+      x' <- generalizeInstance dTy x
+      return (Abs Empty x', [])
+    _ -> error "not implemented"
+  refreshAbs ab \newBs x' -> do
+    Abs bs' UnitE <- applySubst (b @> SubstVal x') $ Abs bs UnitE
```

Comment:
This looks quadratic. Perhaps carry the SubstFrag in parameters and apply it right before you actually use a binder?

Reply:
Good idea